docs/metrics-data-model.md

```diff
@@ -0,0 +1,108 @@
+# k6 Time Series
+
+We want to introduce the time series concept to k6 for getting the benefit of efficiently identifying the combination between a metric and the relative tags.
+
+A summary of the benefits it brings across the platform:
+
+* It allows executing faster lookups and comparisons by a single value, the ID (hash) of the series.
+* It reduces the storage by allocating just one time a possible combination (the time series) and referencing it in all other places. 
+* It allows aggregating data by time series.
+* It enables easier detection of high cardinality issues just by counting the entire unique data set of generated time series.
+* It enables time series based outputs to simplify the integration's logic (e.g. Prometheus).
+
+## Data model
+
+#### Metrics and Series
+
+* Metric (name, type)
+* Time Series (metric, tags)
+* Sample (timestamp, value, time series ref)
+* Tag - (aka [k6/1831](https://github.com/grafana/k6/issues/1831))
+
+```go
+// This is the current k6 Metric polished from the Thresholds dependencies.
+type Metric struct {
+	name     string
+	type     MetricType
+	contains ValueType
+}
+
+type Tag struct {
+    Key, Value string
+}
+
+type TimeSeries struct {
+	id uint64 // hash(metric+tags)
+	metricName string
+	tags   []Tag
+	keys map[string]struct{}
+}
+
+type Sample struct {
+        TimeSeries *TimeSeries
+        Timestamp uint64
+        Value float64
+}
+```
+
+> **Note**: Tag and Metric without Sink+Threholds are represented here just as a general overview of the potential final data model. Tag is to be considered as [k6/1831](https://github.com/grafana/k6/issues/1831) and it isn't expected to be addressed in the same iteration.
+
+#### IDs
+
+The series IDs are `uint64` values generated by hashing the metric name and the tags' key-value pairs.
+
+#### Sink
+
+The sinks are implemented by metric types and they keep the series values up to date:
+
+* Counter:  a monotonic increasing value
+* Gauge: the latest value
+* SparseHistogram: counters per dynamic ranges (buckets) - https://grafana.com/blog/2021/11/03/how-sparse-histograms-can-improve-efficiency-precision-and-mergeability-in-prometheus-tsdb
+
+## Storage
+
+#### Time Series database (aka tsdb)
+
+We need to store all the time series generated from `k6` run during the execution so the other components (mostly the outputs) could query the storage for getting the value of the time series. It's expected to be in-memory and it should be concurrent-safe.
+
+```go
+type Repository interface {
+    InsertSeries(TimeSeries) error
+    GetSeriesByID(id uint64) (*TimeSeries, error)
+}
```

Comment:
Hmm :thinking: You mention that this should be "concurrent-safe", by which I presume you mean that it's safe to call these methods across multiple goroutines? 

If so, then there is a slight problem here, a potential data race more specifically. I see it in the code below as well:
```go
series, err = o.tsdb.GetSeriesByID(id)
if err != nil {
	if !errors.Is(err, tsdb.ErrSeriesNotFound) {
		return 0, err
	}
	series = o.k6MetricToSeries(sample.Metric, tset)
	err = o.tsdb.InsertSeries(series)
	// ...
}
```

Notice that if 2 concurrent goroutines encounter a new time series at the same time, they will both receive a `tsdb.ErrSeriesNotFound` error from `tsdb.GetSeriesByID()`. Then they will both try to call `tsdb.InsertSeries(series)`, but only one of them will succeed.

Instead, I suggest you expose only one method, something like this:
```go
type Repository interface {
    GetOrCreateSeries(id uint64, tags []Tag, metric *Metric) *TimeSeries
    // other methods - GetByID, GetByMetric, GetAll? not sure what else would be required here
}
```

It can do the locking internally and either return the already existing `TimeSeries` or create a new one without any races :man_shrugging: Alternatively, it can probably be the place where hashes are calculated, so you don't need to pass them :thinking: Something like this:
type Repository interface {
    GetOrCreateSeries(tags []Tag, metric *Metric) *TimeSeries // you can query the hash (ID) from the returned *TimeSeries
    // Some other potentially useful methods, though I am not sure if they are needed:
    GetSeriesWithHash(id uint64) (*TimeSeries, error)
    GetSeriesWithMetric(*Metric) []*TimeSeries 
}

Reply:
You're right, I thought about it with the initial reason of having them generated in a central place. Going to update it.